### ISSUE TITLE
Add subword tokenization and support training with subword units

### ISSUE THREAD

**Issue by PiotrCzapla** [n-waves/ulmfit-multilingual#3]

**Comment by NirantK** [comment]
Should we consider using Google's [sentencepiece](https://github.com/google/sentencepiece) for sub word tokenization? 

Have you tried it with any European (i.e. non-English) languages @PiotrCzapla ? Or seen someone try it with Asian or Indic languages?

**Comment by dust0x** [comment]
Hi Nirant, I used it on German for the GermEval experiments we ran. Piotr has also used it on both German and Polish. I personally did not see any gains from sub word tokenization, in fact the results were worse.

There may be benefits on languages such as Japanese that do not have any marked separation between characters (spaces and such). I do not know of anyone who may have used on Indic languages. How did you tokenize the Hindi dataset?

**Comment by dust0x** [comment]
@PiotrCzapla it looks like the [current tokenization implementation](https://github.com/n-waves/ulmfit-multilingual/blob/master/ulmfit/postprocess_wikitext.py#L17) is a simple split on space character.

What exactly is the fastai v1 way of tokenizing?

**Comment by NirantK** [comment]
@sebastianruder thanks for pointing to MosesTokenizer. 

IIRC MosesTokenizer does word level tokenization. Does that mean our models trained on this will still treat OOV words as `<unk>`? 

Is there a way to have a better embedding than `<unk>`?

**Comment by sebastianruder** [comment]
I can't think of a good way other than training with the `<unk>` embedding as we already do. To address this, I think it's important that we train separate models both with word and subword tokenization, so that we can compare them for different languages.

**Comment by dust0x** [comment]
Yes, that's the only logical way. I'll write sentencepiece into the script.

**Comment by NirantK** [comment]
@dust0x maybe add `sentencepiece` in a way that we still default to `MosesTokenizer`? 

Also, would you mind trying the `sentencepiece` and `MosesTokenzier` both with QRNN on German? I believe that your earlier was work with AWD-LSTM?

**Comment by dust0x** [comment]
Yes, that'll be the idea. I'll also have to see about train_clas.py which directly calls on `MosesTokenizer` at the moment.

**Comment by NirantK** [comment]
Also, quick note - I tried Hindi-QRNN with no vocab limits with `MosesTokenizer`. I had earlier used a spaCy multi-lingual tokenizer. It is a 10 points jump on test perplexity from previous SoTA 

Previous: ~47 [here](https://github.com/NirantK/hindi2vec) 
Now: ~36 with ulmfit at [this commit](https://github.com/n-waves/ulmfit-multilingual/commit/70680c21ce808de6185609b8c8be353b87ab4c38)

**Comment by PiotrCzapla** [comment]
> I personally did not see any gains from sub word tokenization, in fact the results were worse.
@dust0x  but you have quite clever pre processing and you got a bit better results than my sentence piece model without any preprocessing. Ie leaving urls, usernames and emjois in. 
So I think sentence piece does provide some benefits over word tokens.

> we should just use the MosesTokenizer instead.
@sebastianruder ,  ok, but can you share some thoughts about this decision. It can be on priv.

> Is there a way to have a better embedding than <unk>?
sentence piece will make the unknown tokens out of letters of word fragments, you don't have any <unk> then.

> I'll write sentencepiece into the script.
@dust0x  awesome ! btw. Marcin prepared some code to run sentence piece directly from python with some optimization to process text in parallel and to use Sentence Piece "Subword regularization" . Subword regularization with TTA was something I had high hopes but did not work that well in the end, but maybe we haven't gave it enough love :),  have a look here:  https://github.com/n-waves/fastai/compare/poleval/task3...n-waves:subwordreg?expand=1
The code isn't polished yet but I think you might find it useful.

>Also, quick note - I tried Hindi-QRNN with no vocab limits with MosesTokenizer. I had earlier used a >spaCy multi-lingual tokenizer. It is a 10 points improvement on test perplexity from previous SoTA
Awesome!

**Comment by sebastianruder** [comment]
@dust0x, let us know how you're getting on with adding sentencepiece. Running it directly from Python has @PiotrCzapla pointed out would be great.
Once this is in place, @eisenjulian wanted to try to see if we can train a multilingual model by concatenating multiple Wikipedias and training a joint LM on them.

**Comment by eisenjulian** [comment]
Hi everybody, thanks for the invite. I have been working with @sebastianruder for quite some time in different multi-lingual classification datasets and I am eager to join forces. 

Maybe it's useful, but [Here](https://github.com/eisenjulian/fastai/compare/sentencepiece?expand=1) I have a branch on my fastai fork that uses the sentencepiece python bindings, it works with the latest iteration of v1, or at least it did a few days ago when I tried last :) It doesn't preprocess with Moses as BERT does, but I think that could be added easily as a `rule`. 

I see there's a lot going on and I don't want to overlap with anybody else's work, so please let me know if you think that branch is worth merging, or I will skip directly into bilingual model training.

**Comment by dust0x** [comment]
Hi @eisenjulian, I had only started writing the code a little before your comment. Your implementation should be perfect, I think. Maybe you can create a branch from the n-waves/fastai and test this with the ulmfit scripts we're using.

I can help you update the ulmfit scripts.

**Comment by dust0x** [comment]
@sebastianruder just started a PR with some initial code. Since Julian just shared his code, I think it's easiest if I write around it with a few small changes.

Another thing I am thinking of doing in the same PR is to do away with the postprocessing script mainly because sentencepiece does not need most of what it does and (please correct me if I am wring here) all of it can be pretty much implemented within the create script.

**Comment by sebastianruder** [comment]
Great!
Good point re the post-processing script. I've created it originally to create small and large datasets to benchmark LMs in different languages (similar to the WikiText-2 and WikiText-103 datasets). At this point, there are already a couple of small LM datasets in different languages available, so I don't think we need to generate these anymore.

**Comment by PiotrCzapla** [comment]
@eisenjulian your sentence piece code looks nice, and it is self-contained so it is a perfect fit for `ulmfit-mutlingual/fastai_contrib` . All classes there will eventually end up in fastai, but we have a separate module for the process of development to make it easy to upgrade to new version of fastai. `n-waves/fastai` should be used mostly for bug fixing and pushing PR to `fastai/fastai:master`.

**Comment by sebastianruder** [comment]
On another note: Do we want to tokenize with Moses before running sentence piece? XNLI is already tokenized with Moses, so it might be beneficial to do that to do better on that dataset.

**Comment by dust0x** [comment]
@sebastianruder yes, that us another thing I considered. If I understand correctly, Moses does some initial clean up of the text which should help. Juilian's spm class has the functionality of adding additional `rules` and as he says Moses could probably be integrated into it as a rule.

@PiotrCzapla, that's my bad actually. I had meant to say `ulmfit-mutlingual/fastai_contrib` in my comment (as I have in the PR).

**Comment by sebastianruder** [comment]
@dust0x, cool. In that case, in order to keep things simple and so that we don't have too many versions of the same dataset lying around, I'd suggest that we use Moses in every case, so for both word level tokenization and before sentence piece.

**Comment by eisenjulian** [comment]
Awesome, will move the implementation to the `sentencepiece` branch. @dust0x I have tested the tokenizer in the following way
```python
# Trains or loads a model from disk
sp = get_sentencepiece(path, dataset)
# Uses pre-built Vocab and Tokenizer instances
lm_data = TextLMDataBunch.from_csv(path, 'train.csv', **sp) 
```

**Comment by eisenjulian** [comment]
Sorry if this has been brought up before, or if there is an obvious answer I am not seeing: 
On a theoretical note, I am curious if it's fair to compare perplexity results when switching between different tokenization strategies. I understand that the perplexity will depend heavily on the specific tokens and although it can be normalized according to vocabulary size, it's far from perfectly comparable. Shouldn't the deciding factor be downstream task performance?

**Comment by PiotrCzapla** [comment]
> I am curious if it's fair to compare perplexity results when switching between different tokenization strategies.
You can convert perplexity between sentence piece and regular tokenization providing that you ensure that number of out of vocabulary words stays the same.  We have code for that in poleval repo.
and here is [why this work (section 3.2)](https://arxiv.org/pdf/1810.10222.pdf)

> Shouldn't the deciding factor be downstream task performance?
It is definitely the safest, but it will be hard to experiment. so let's try to use perplexity where we can. We can benchmark different models against let say WikiText-2 with a fixed number of oov tokens. I think this is what @sebastianruder wanted to do have a look here:
https://github.com/n-waves/ulmfit-multilingual/blob/master/ulmfit/pretrain_lm.py#L126-L134

**Comment by sebastianruder** [comment]
Yeah, perplexity estimates would be nice, but as you say aren't comparable between different word level tokenization strategies or between different vocabulary sizes. Perplexity for regular tokenization is comparable with sentence piece tokenization, but that is given that we train on the full vocabulary, which we don't at the moment. 
In the end, I think we should report _some_ perplexity numbers to show that our models are reasonably good language models, but I think the final performance metric, the accuracy on the downstream task, will be the main metric to evaluate on.

**Comment by eisenjulian** [comment]
@PiotrCzapla perfect, as I commented in the other thread, you're right, we should use the same model always. Good catch! Should be close this issue? 

I am curious about one more thing. I saw that some changes were made at the wikipedia preprocessing stage, and I want to retrain the bi-lingual models with that, do you know the command you are using for pre-training with sentencepiece? Mainly, I wonder if we should use the version that removes infrequent tokens or the full one, given that sentencepiece can account for that regardless. Thank you and happy new year!

**Comment by PiotrCzapla** [comment]
@eisenjulian Hi, Happy new year.
I've changed the way we create the wikipedia data sets to be able to split them by articles so that LM can learn long dependencies. Short articles break the downstream tasks.

I haven't had a chance to train BiLM after the changes as I'm down to 11GB :/. But here is how I would do that:
python -m ulmfit lm --dataset-path data/wiki/wikitext-103 --bidir=True --qrnn=True --nl=4 --tokenizer=sp --name 'nl4' --bs 120 --cuda-id 0  -  train 10 --drop-mult=0 --bs 40

**Comment by dust0x** [comment]
Happy new year everyone.

@PiotrCzapla, thanks for undertaking the work in code refactoring, it looks great! Everything seems to be falling into place. Time to push for XNLI now :)

Also, should this issue be closed now that the refactored master supports subword tokenization at all stages?

**Comment by PiotrCzapla** [comment]
It  does not work for xnli yet.
